src/custom/conversions.rs

```diff
@@ -0,0 +1,463 @@
+                )
+                .ok_or(anyhow!("TODO"))?;
+
+            // We have to check if the conversion will work, so we
+            // make the maximum int and convert to a float, then compare
+            // with the function input.
+            let int_max = if signed {
+                int_ty.const_int(int_max_value_s, false)
+            } else {
+                int_ty.const_int(int_max_value_u, false)
+            };
+
+            let flt_int_max = if signed {
+                let abs_name = &format!("llvm.abs.i{}", width);
+                let abs_intr = Intrinsic::find(abs_name).expect("Couldn't find int abs intrinsic");
+                let abs = abs_intr
+                    .get_declaration(ctx.get_current_module(), &[int_ty.as_basic_type_enum()])
+                    .unwrap();
+
+                let const_false = ctx
+                    .iw_context()
+                    .bool_type()
+                    .const_zero()
+                    .as_basic_value_enum();
+                let abs_call = ctx
+                    .builder()
+                    .build_call(abs, &[int_max.into(), const_false.into()], "max_int_pos")?
+                    .as_any_value_enum()
+                    .into_int_value();
+
+                ctx.builder().build_signed_int_to_float(
+                    abs_call,
+                    ctx.iw_context().f64_type(),
+                    "max_flt_pos",
+                )
+            } else {
+                ctx.builder().build_unsigned_int_to_float(
+                    int_max,
+                    ctx.iw_context().f64_type(),
+                    "max_flt",
+                )
+            }?;
+
+            let fabs_call = ctx
+                .builder()
+                .build_call(fabs, &[arg.into()], "flt_pos")?
+                .as_any_value_enum()
+                .into_float_value();
+
+            // TODO: Test this with converting INT_MAX to float to see if we need a cheeky error margin
+            let success = ctx.builder().build_float_compare(
+                FloatPredicate::OLE,
+                fabs_call,
+                flt_int_max,
+                "conversion_valid",
+            )?;
+            let success = ctx.builder().build_int_z_extend(
+                success,
+                ctx.iw_context().i32_type(),
+                "conversion_valid_i32",
+            )?;
+
+            // Perform the conversion unconditionally, which will result
+            // in a poison value if the input was too large. We will
+            // decide whether we return it based on the result of our
+            // earlier check.
+            let trunc_result = if signed {
+                ctx.builder()
+                    .build_float_to_signed_int(arg.into_float_value(), int_ty, "")
+            } else {
+                ctx.builder()
+                    .build_float_to_unsigned_int(arg.into_float_value(), int_ty, "")
+            }?
+            .as_basic_value_enum();
```

Comment:
Might be better to branch on `success` instead of always doing the cast?

Reply:
I don't think this is too bad... I've addressed all the comments except this one - take a look and let me know whether you think it's essential.